core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -315,6 +315,13 @@ private[chisel3] trait HasId extends chisel3.InstanceId {
         case (Some(c), _) => refName(c)
         case (None, d: Data) if d.topBindingOpt == Some(CrossModuleBinding) => _ref.get.localName
         case (None, _: MemBase[_]) => _ref.get.localName
+        case (None, _) if _ref.isDefined => {
+          // Support instance names for HasIds that don't have a _parent set yet, but do have a _ref set.
+          // This allows most HasIds to be named in atModuleBodyEnd, for example.
+          // However, this is only valid if the _ref's fullName doesn't depend on the component name.
+          // Notably, this is not valid for ports of instances, and other constructs that need the _parent's name.
```

Comment:
This comment is a little bit concerning, what happens if you use `.toRelativeTarget` on a port of a submodule?

Reply:
I think I may have made this comment sound too scary, without thinking deeply enough about it. I just pushed up a commit with a submodule, is this what you had in mind? [68a31cb](https://github.com/chipsalliance/chisel/pull/3623/commits/68a31cb4a8efd6d6027344ca9da7762d03b03d40)

I think we actually are fine here. If we're trying to name a port of a submodule, the _parent component of the port will always be set, so we wouldn't come through this code path anyway. The only time we should come here is if we're naming a port of the currently-being-elaborated module, in which case the localName is safe to use (I also added a test demonstrating that).

Does this make sense? Maybe I just need to amend the comment to indicate that in fact any time we hit this code path, it's safe to use a localName.

Reply:
FYI I've pushed another commit that updates the comment based on what I just said here.

Reply:
Yep makes sense, thank you for the test and clarification!